### PR TITLE
Implement azurerm_automation_certificate

### PR DESCRIPTION
**Pull request by dowlingw**
Implement `azurerm_automation_certificate` as requested in #3881 

**Overview of changes:**
* Added CertificateClient to `automation` client
* Implemented new resource type `azurerm_automation_certificate`
* Resource documentation for website

**Notes**
This is my first attempt at Terraform provider code, any feedback about how this could be done better or differently would be appreciated.

I've elected to set `Computed: true` on the `is_exportable` and `thumbprint` resource properties, even though these can be set on the Create/Update calls - these are extracted from the provided certificate and returned in any REST responses. I've validated this for the `thumbprint` property but not for `is_exportable`. I can't imagine anyone would need to doubly specify this.

### CODE REVIEW

File: azurerm/resource_arm_automation_certificate.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"account_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: Could we validate this with at least `validate.NoEmptyStrings`?

File: azurerm/resource_arm_automation_certificate.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: Could we validate this with at least `validate.NoEmptyStrings`?

File: azurerm/resource_arm_automation_certificate.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"base64": {
+				Type:      schema.TypeString,
+				Required:  true,
+				ForceNew:  true,
+				Sensitive: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: could we validate this is base64?

File: azurerm/resource_arm_automation_certificate.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"base64": {
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: would a different name for this property make more sense? like `certificate_base64`?

File: azurerm/resource_arm_automation_certificate.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"account_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"name": {
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: Could we move name to the top of the schema for consistency?

File: azurerm/resource_arm_automation_certificate.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+			},
+
+			"base64": {
+				Type:      schema.TypeString,
+				Required:  true,
+				ForceNew:  true,
+				Sensitive: true,
+			},
+
+			"is_exportable": {
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: should this not be computed as we are setting it?

File: azurerm/resource_arm_automation_certificate.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+func resourceArmAutomationCertificateCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).Automation.CertificateClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for AzureRM Automation Certificate creation.")
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: could we use `resourceGroup` instead?
```suggestion
	resourceGroup := d.Get("resource_group_name").(string)
```

File: azurerm/resource_arm_automation_certificate.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	accName := d.Get("account_name").(string)
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: and accountName?
```suggestion
	accountName := d.Get("account_name").(string)
```

File: azurerm/resource_arm_automation_certificate.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+	description := d.Get("description").(string)
+
+	parameters := automation.CertificateUpdateParameters{
+		CertificateUpdateProperties: &automation.CertificateUpdateProperties{
+			Description: &description,
+		},
+		Name: &name,
+	}
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: I am going to guess the missing properties here dont' exist in the update struct?

File: website/docs/r/automation_certificate.html.markdown [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The Automation Certificate ID.
+
+* `is_exportable` - The is exportable flag of the certificate.
+
+* `thumbprint` - The thumbprint for the certificate.
```
Comment (katbyte) [inline website/docs/r/automation_certificate.html.markdown]: you are setting these in the create function?

File: website/docs/r/automation_certificate.html.markdown [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+* `name` - (Required) Specifies the name of the Certificate. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the resource group in which the Certificate is created. Changing this forces a new resource to be created.
+
+* `account_name` - (Required) The name of the automation account in which the Certificate is created. Changing this forces a new resource to be created.
+
+* `base64` - (Required) Base64 encoded value of the certificate.
+
+* `description` -  (Optional) The description associated with this Automation Certificate.
```
Comment (katbyte) [inline website/docs/r/automation_certificate.html.markdown]: phrasing might be better as 
```suggestion
* `description` -  (Optional) The description of this Automation Certificate.
```

File: azurerm/resource_arm_automation_certificate.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+	base64 := d.Get("base64").(string)
+	isExportable := d.Get("is_exportable").(bool)
+	thumbprint := d.Get("thumbprint").(string)
+	description := d.Get("description").(string)
+
+	parameters := automation.CertificateCreateOrUpdateParameters{
+		CertificateCreateOrUpdateProperties: &automation.CertificateCreateOrUpdateProperties{
+			Base64Value:  &base64,
+			IsExportable: &isExportable,
+			Thumbprint:   &thumbprint,
```
Comment (katbyte) [inline azurerm/resource_arm_automation_certificate.go]: these are computed not optional?

File: azurerm/resource_arm_automation_certificate.go (hunk 4 above)
Comment (dowlingw) [reply azurerm/resource_arm_automation_certificate.go]: I did have a quick look at existing resources in the provider and aligned to those cases using `base64`. I figure in context of the resource declaration for `certificate` it's probably clear in context?

File: azurerm/resource_arm_automation_certificate.go (hunk 9 above)
Comment (dowlingw) [reply azurerm/resource_arm_automation_certificate.go]: Correct, as per: https://docs.microsoft.com/en-us/rest/api/automation/certificate/update

As a side note, the `PUT` (CreateOrUpdate) call does imply support for updating other properties, but I haven't actually validated whether those changes would be accepted by implementation.

Not sure whether this is being overly conservative, but thinking about the typical use cases - the other fields all relate to certificate material, which are all use cases around destroy/replacing the resource anyway.

File: azurerm/resource_arm_automation_certificate.go (hunk 4 above)
Comment (dowlingw) [reply azurerm/resource_arm_automation_certificate.go]: Marking this as resolved as I think this is blocking having this looked at - and I think the argument that this aligns with existing `base64` certificate parameters is reasonable.

Definitely happy for more conversation around this however :)

**Comment by dowlingw** [comment]
I think this should be ready to go, please have a look and let me know if you've got any further feedback <3

**Comment by tombuildsstuff** [comment]
Tests pass:

```
$ acctests azurerm TestAccAzureRMAutomationCertificate_
=== RUN   TestAccAzureRMAutomationCertificate_basic
=== PAUSE TestAccAzureRMAutomationCertificate_basic
=== RUN   TestAccAzureRMAutomationCertificate_requiresImport
--- SKIP: TestAccAzureRMAutomationCertificate_requiresImport (0.00s)
    resource_arm_automation_certificate_test.go:44: Skipping since resources aren't required to be imported
=== RUN   TestAccAzureRMAutomationCertificate_complete
=== PAUSE TestAccAzureRMAutomationCertificate_complete
=== RUN   TestAccAzureRMAutomationCertificate_update
--- PASS: TestAccAzureRMAutomationCertificate_update (201.07s)
=== CONT  TestAccAzureRMAutomationCertificate_basic
--- PASS: TestAccAzureRMAutomationCertificate_basic (109.49s)
=== CONT  TestAccAzureRMAutomationCertificate_complete
--- PASS: TestAccAzureRMAutomationCertificate_complete (107.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	418.130s
```

**Comment by dowlingw** [comment]
No problems whatsoever, thanks for sorting those out - cheers :)

**Pull request merged**